File: Card Transition Demo/Controller/ViewController.m
//
//  ViewController.m
//  Card Transition Demo
//
//  Created by Mingfei Huang on 11/4/18.
//  Copyright © 2018 Mingfei Huang. All rights reserved.
//

#import "ViewController.h"

#import <YYKit/YYKit.h>

#import "KAKAnimation.h"

@interface ListVC () <KAKAnimationPresentingViewController>

@property (strong, nonatomic) KAKAnimation *currentAnimation;

@end

@interface DetailVC () <KAKAnimationPresentedViewController, UIScrollViewDelegate>

@property (strong, nonatomic) UIScrollView *scrollView;

@end

@implementation ListVC

- (void)viewDidLoad {
    [super viewDidLoad];
    self.view.backgroundColor = [UIColor lightGrayColor];
    [self setupCardView];
}

- (void)setupCardView
{
    NSString *imageUrl = @"https://source.unsplash.com/random";
    
    UIImageView *cardView = [UIImageView new];
    [cardView setImageWithURL:[NSURL URLWithString:imageUrl] options:YYWebImageOptionRefreshImageCache];
    cardView.width = cardView.height = 200;
    cardView.contentMode = UIViewContentModeScaleAspectFill;
    cardView.layer.cornerRadius = 8;
    cardView.clipsToBounds = YES;
    cardView.userInteractionEnabled = YES;
    cardView.center = self.view.center;
    cardView.autoresizingMask = UIViewAutoresizingFlexibleTopMargin | UIViewAutoresizingFlexibleLeftMargin | UIViewAutoresizingFlexibleRightMargin | UIViewAutoresizingFlexibleBottomMargin;
    
    [self.view addSubview:cardView];
    
    __weak typeof(cardView) weak_cardView = cardView;
    __weak typeof(self) weak_self = self;
    
    [cardView addGestureRecognizer:[[UITapGestureRecognizer alloc] initWithActionBlock:^(id  _Nonnull sender) {
        if (weak_self.view.window) {
            DetailVC *detailVC = [DetailVC instanceWithLaunchingAnimation:weak_self.currentAnimation];
            weak_self.currentAnimation = [KAKAnimation instanceWithTriggerView:weak_cardView
                                                      presentingViewController:weak_self
                                                       presentedViewController:detailVC];
            detailVC.anime = weak_self.currentAnimation;
            detailVC.transitioningDelegate = weak_self.currentAnimation;
            [weak_self presentViewController:detailVC animated:YES completion:nil];
        } else {
            [weak_self.currentAnimation dismissInteractively:NO];
        }
    }]];
}

- (void)completeDismissAnimation:(UIView *)triggerView
{
    triggerView.autoresizingMask = UIViewAutoresizingFlexibleTopMargin | UIViewAutoresizingFlexibleLeftMargin | UIViewAutoresizingFlexibleRightMargin | UIViewAutoresizingFlexibleBottomMargin;
}

@end

@implementation DetailVC

+ (instancetype)instanceWithLaunchingAnimation:(KAKAnimation *)anime
{
    DetailVC *inst = self.new;
    inst.anime = anime;
    return inst;
}

- (void)viewDidLoad
{
    [super viewDidLoad];
    self.view.backgroundColor = [UIColor whiteColor];
    
    UIScrollView *scrollView = [UIScrollView new];
    scrollView.frame = self.view.bounds;
    scrollView.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
    scrollView.bounces = NO;
    self.scrollView = scrollView;
    [self.view addSubview:scrollView];
    
    UIPanGestureRecognizer *gest = [[UIPanGestureRecognizer alloc] initWithTarget:self action:@selector(handlePanGesture:)];
    [self.view addGestureRecognizer:gest];
}

- (void)preparePresentAnimation:(UIView *)triggerView
{
    [self.scrollView addSubview:triggerView];
}

- (void)performPresentAnimation:(UIView *)triggerView
{
    triggerView.frame = CGRectMake(0, 0, self.view.width, 400);
    triggerView.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleBottomMargin | UIViewAutoresizingFlexibleLeftMargin | UIViewAutoresizingFlexibleRightMargin;
}

- (void)handlePanGesture:(UIPanGestureRecognizer *)gesture
{
    CGFloat percentage = [gesture translationInView:self.view].y / 200;
    if (gesture.state == UIGestureRecognizerStateBegan) {
        if (percentage > 0) {
            [self.anime dismissInteractively:YES];
        } else {
            gesture.enabled = NO;
            gesture.enabled = YES;
        }
    } else if (gesture.state == UIGestureRecognizerStateChanged) {
        [self.anime updateInteractiveTransition:percentage];
    } else if (gesture.state == UIGestureRecognizerStateEnded) {
        if (percentage > 0.5) {
            [self.anime finishInteractiveTransition];
        } else {
            [self.anime cancelInteractiveTransition];
        }
    }
}

@end

File: Card Transition Demo/Controller/ViewController.swift
//
//  ViewController.swift
//  Card Transition Demo
//
//  Created by Mingfei Huang on 11/4/18.
//  Copyright © 2018 Mingfei Huang. All rights reserved.
//

import YYKit

class ListVC: UIViewController {
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        self.view.backgroundColor = UIColor.lightGray
        self.setupCardView()
    }
    
    func setupCardView() {
        let imageUrl = "https://source.unsplash.com/random"
        let cardView = UIImageView(frame: CGRect(x: 0, y: 0, width: 200, height: 200))
        cardView.setImageWith(URL(string: imageUrl), options: YYWebImageOptions.refreshImageCache)
        cardView.contentMode = UIView.ContentMode.scaleAspectFill
        cardView.layer.cornerRadius = 8
        cardView.clipsToBounds = true
        cardView.isUserInteractionEnabled = true
        cardView.center = self.view.center
        cardView.autoresizingMask = [
            .flexibleTopMargin,
            .flexibleLeftMargin,
            .flexibleRightMargin,
            .flexibleBottomMargin
        ]
        self.view.addSubview(cardView)
        
        var currentAnimation: KAKAnimation!
        cardView.addGestureRecognizer(UITapGestureRecognizer.init(actionBlock: { [unowned self, cardView] (_) in
            if (self.view.window != nil) {
                let detailVC = DetailVC(nibName: nil, bundle: nil)
                currentAnimation = KAKAnimation(triggerView: cardView,
                                         presentingViewController: self,
                                         presentedViewController: detailVC)
                detailVC.anime = currentAnimation
                detailVC.transitioningDelegate = currentAnimation
                self.present(detailVC, animated: true, completion: nil)
            } else {
                currentAnimation.dismiss(interactively: false)
            }
        }))
    }
    
}

extension ListVC: KAKAnimationPresentingViewController {
    func completeDismissAnimation(_ triggerView: UIView) {
        triggerView.autoresizingMask = [
            .flexibleTopMargin,
            .flexibleLeftMargin,
            .flexibleRightMargin,
            .flexibleBottomMargin
        ]
    }
}

class DetailVC: UIViewController {
    var anime: KAKAnimation!
    var scrollView: UIScrollView!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        self.view.backgroundColor = UIColor.white
        self.setupScrollView()
        self.setupDismissGesture()
    }
    
    func setupScrollView() {
        self.scrollView = UIScrollView(frame: self.view.bounds)
        self.scrollView.autoresizingMask = [.flexibleWidth, .flexibleHeight]
        self.scrollView.bounces = false
        self.view.addSubview(scrollView)
    }
    
    func setupDismissGesture() {
        self.view.addGestureRecognizer(UIPanGestureRecognizer(target: self, action: #selector(handlePanGesture(gesture:))))
    }
    
    @objc
    func handlePanGesture(gesture: UIPanGestureRecognizer) {
        let percentage = gesture.translation(in: self.view).y / 200
        if (gesture.state == .began) {
            if (percentage > 0) {
                self.anime.dismiss(interactively: true)
            } else {
                gesture.isEnabled = false
                gesture.isEnabled = true
            }
        } else if (gesture.state == .changed) {
            self.anime.update(percentage)
        } else if (gesture.state == .ended || gesture.state == .cancelled) {
            if (percentage > 0.5) {
                self.anime.finish()
            } else {
                self.anime.cancel()
            }
        }
    }
}

extension DetailVC: KAKAnimationPresentedViewController {
    func preparePresentAnimation(_ triggerView: UIView) {
        self.scrollView.addSubview(triggerView)
    }
    
    func performPresentAnimation(_ triggerView: UIView) {
        triggerView.frame = CGRect(x: 0, y: 0, width: self.view.width, height: 400)
        triggerView.autoresizingMask = [
            .flexibleWidth,
            .flexibleLeftMargin,
            .flexibleRightMargin,
            .flexibleBottomMargin
        ]
    }
}
